### DISCUSSION TITLE
How is it possible that fkyaml is about 2 times slower than nlohmann json?

### DISCUSSION THREAD

**Discussion by sndth** [fktn-k/fkYAML#446] [Q&A]
Although `YAML` is much simpler in its structure, and the file size is `~200KB` smaller after converting from `JSON` to `YAML`, the parsing time of the document is significantly slower compared to the `nlohmann::json`. Here’s the benchmark using [nanobench](https://github.com/martinus/nanobench) library:

### Debug (x86)
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|      645,453,900.00 |                1.55 |    2.7% |      7.29 | `nlohmann::json`
|    1,272,145,000.00 |                0.79 |    0.8% |     14.02 | `fkyaml`

### Debug (x64)
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|      281,163,900.00 |                3.56 |    2.7% |      3.31 | `nlohmann::json`
|      713,596,700.00 |                1.40 |    2.8% |      7.82 | `fkyaml`

### Release (x86)
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|       15,199,700.00 |               65.79 |   10.0% |      0.18 | `nlohmann::json`
|       37,735,000.00 |               26.50 |    4.5% |      0.46 | `fkyaml`

### Release (x64)
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|       17,028,300.00 |               58.73 |   13.1% |      0.19 | `nlohmann::json`
|       45,173,100.00 |               22.14 |   17.9% |      0.51 | `fkyaml`

## Example test code to reproduce:
test.yaml
```yaml
a: b
```
test.json
```json
{ "a": "b" }
```
<hr>

```cpp
#include <fstream>
#include "json.hpp"
#include "node.hpp"

#define ANKERL_NANOBENCH_IMPLEMENT

#include "nanobench.h"

int
main()
{
  nlohmann::json j;
  ankerl::nanobench::Bench().run("nlohmann::json", [&] {
    std::ifstream ifs("test.json");
    j = nlohmann::json::parse(ifs);
    ankerl::nanobench::doNotOptimizeAway(j);
  });

  fkyaml::node y;
  ankerl::nanobench::Bench().run("fkyaml", [&] {
    std::ifstream ifs("test.yaml");
    y = fkyaml::node::deserialize(ifs);
    ankerl::nanobench::doNotOptimizeAway(y);
  });
}
```
### Result on Debug x86:
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          138,862.50 |            7,201.37 |    1.9% |      0.01 | `nlohmann::json`
|          206,460.00 |            4,843.55 |    3.1% |      0.01 | `fkyaml`

**Comment by fktn-k** [comment]
@sndth   
Thanks for the question.  
I'd like to investigate the reason(s) later, but let me clarify this first:  

YAML “looks“ much simpler than JSON, but it ISN'T simple as such or rather much more complicated in terms of its specification, in other words, requirements for parser implementation.  

That affects the performance as the benchmark scores in the README shows.  
So, if you want to compare performance with different libraries, it should be better to consider input formats as well.

Here is a case where YAML is more complicated than JSON, taking the snippets you shared above.  
(This is long. you can skip it)  

A simple plain (unquoted) scalar, for example, has many more restrictions for allowed characters than quoted ones, and  a not-allowed character such as `: `  also a terminator of the plain scalar, while a quoted scalar (the only way to express a string in JSON) is reasonably expected to be ended with a quotation mark and that's much simpler and strait forward.  
Actually, implementation for parsing plain scalars is at least twice as complicated as that for quoted ones.

**Reply by fktn-k** [reply]
@sndth   
I benchmarked the parse performance with [nlohmann/json(v3.11.3)](https://github.com/nlohmann/json/releases/tag/v3.11.3) and [fkYAML(v0.4.0)](https://github.com/fktn-k/fkYAML/releases/tag/v0.4.0).  
I created a simple CMake project to reproduce your results. ([bench_nlohmann_json_and_fkyaml.zip](https://github.com/user-attachments/files/18216913/bench_nlohmann_json_and_fkyaml.zip))   

The results show that fkYAML is at most 1.3x slower than nlohmann/json (not "about 2 times slower" though).  
I know there exists much room for optimization in the deserialization feature, especially in the basic_deserializer class implementation which is not well organized and does too many things.  
Its complexity is sometimes too much depending on input contents, which surely slows down the parse speed.  
If you find some bottleneck and know how to fix it, I'm pleased to accept a pull request.  

Anyway, here are the results I've got locally.  
I used a release build (x64) version of bench_nlohmann_json_and_fkyaml to get them (CPU: AMD Ryzen 7 5800H \@3.20GHz).  

## GCC 11.4.0 on Ubuntu 22.04LTS (WSL2)
* nlohmann/json: `test.json`
* fkYAML: `test.yaml`

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|            4,382.18 |          228,196.83 |    0.4% |      0.01 | `nlohmann::json`
|            5,728.79 |          174,556.87 |    0.6% |      0.01 | `fkyaml`

* nlohmann/json: `test.json`
* fkYAML: `test.json`

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|            4,436.16 |          225,419.97 |    0.7% |      0.01 | `nlohmann::json`
|            5,840.84 |          171,208.32 |    0.6% |      0.01 | `fkyaml`

## MSVC 19.29.30157.0 on Windows 10
* nlohmann/json: `test.json`
* fkYAML: `test.yaml`

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           48,980.95 |           20,416.10 |    1.3% |      0.01 | `nlohmann::json`
|           53,527.27 |           18,682.07 |    0.4% |      0.01 | `fkyaml`

* nlohmann/json: `test.json`
* fkYAML: `test.json`

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           47,144.00 |           21,211.61 |    0.6% |      0.01 | `nlohmann::json`
|           54,352.63 |           18,398.37 |    0.9% |      0.01 | `fkyaml`

**Reply by sndth** [reply]
Great result! I was also looking to see what else can be improved, and I'll definitely get to it soon. Thank you for the development and improvement of this great library!

**Reply by fktn-k** [reply]
@sndth   
Thanks for your contributions too!  
Feel free to ask anything since this library is under development and YAML itself is incredibly complicated 😢
